src/libraries/System.Data.Common/src/System/Data/DataSet.cs

```diff
@@ -77,6 +78,7 @@ public class DataSet : MarshalByValueComponent, IListSource, IXmlSerializable, I
         /// <summary>
         /// Initializes a new instance of the <see cref='System.Data.DataSet'/> class.
         /// </summary>
+        [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
```

Comment:
What about just creating a blank DataSet is unsafe?

Reply:
This is because it implements ISerializable and the method currently needs RUC

Reply:
@vitek-karas @MichalStrehovsky @marek-safar - thoughts on how we can work around this?

Maybe we just leave the `BinaryFormatter` warnings in the library, and if someone ever uses `DataSet` + serialization, they would get a warning coming from `GetObjectData`?

Reply:
That sounds good. This call should produce the warning: https://github.com/dotnet/runtime/pull/52046/files/4872c212644dbc347a2b12f0296f216d738ac6b6#diff-56389de876889f0a4ee7be70d22a8ea76201069ae8d3f65955205cec022b11dbL307

I'm only worried that basically referencing `ISerializable` and this class (anywhere in the app) will cause the warning. It's not "incorrect" though.

Reply:
@vitek-karas your link doesn't seem to be working. I'm a bit confused here - is the suggestion to put RUC on GetObjectData and suppress the message about inheritance?

Reply:
Sorry - I thought GitHub would produce the right URL - apparently not completely. Anyway, the link was supposed to go to this line: https://github.com/dotnet/runtime/blob/c86d1905c3ad439820b6fd8d9fe50a1690dfb1c1/src/libraries/System.Data.Common/src/System/Data/DataSet.cs#L310

We should not add RUC onto a virtual/interface method - that said `ISerializable` is probably one of the few candidates where we could do that. But I still think we should not. The suggestion is what @eerhardt said - don't suppress the warning in `GetObjectData` which comes from calling the binary formatter. Meaning the `GetObjectData` would have no attributes on it. Only suppress it in libraries build (via XML). This way if the `GetObjectData` ends up in user's application (because trimmer thinks it's needed) - then it would produce a warning about using the binary formatter. It's not great, since the warning would not talk about DataSet directly, but it's a start. If it's becomes a problem we can work on improving it.

Reply:
This is really a point in time issue. The code using BinaryFormatter in DataSet will go away in .NET 7 per https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md

We could potentially choose to accelerate that in .NET 6 per the above schedule - "All first-party dotnet org code bases begin migration". So we could do whatever we're about to do for this in 7 now - I don't know how expensive that would be and whether we have budget for it, but it needs to be done within a ~year anyway.

Reply:
@pgovind - was DataSet on your list of "things to move off of BinaryFormatter in .NET 6"?

According to the obsoletion doc, "The entirety of the ISerializable type. Includes any public or explicitly-implemented GetObjectData methods." will be obsolete in .NET 7. That makes me more comfortable with the plan of just leaving ILLink warnings from `GetObjectData` in the library. It isn't perfect for .NET 6, but I think it is better than marking all the ctors of DataSet as `RequiresUnreferencedCode`.

Reply:
I'm resolving this since I've added this back to suppression file. We can discuss this further once we address this specific comment.